Source/Task/ThreadPool_stl.cpp

```diff
@@ -0,0 +1,130 @@
+#include "pch.h"
+#include "ThreadPool.h"
+
+class ThreadPoolImpl
+{
+public:
+
+    ~ThreadPoolImpl() noexcept
+    {
+        Terminate();
+    }
+
+    HRESULT Initialize(
+        _In_opt_ void* context,
+        _In_ ThreadPoolCallback* callback) noexcept
+    {
+        m_context = context;
+        m_callback = callback;
+
+        uint32_t numThreads = std::thread::hardware_concurrency();
+        if (numThreads == 0)
+        {
+            numThreads = 1;
+        }
+
+        try
+        {
+            while(numThreads != 0)
+            {
+                numThreads--;
+                m_pool.emplace_back(std::thread([this]
+                {
+                    std::unique_lock<std::mutex> lock(m_wakeLock);
+                    while(true)
+                    {
+                        m_wake.wait(lock);
+
+                        if (m_terminate)
+                        {
+                            break;
+                        }
+
+                        if (m_calls != 0)
+                        {
+                            m_calls--;
+
+                            lock.unlock();
+                            m_callback(m_context);
+                            lock.lock();
+                        }
+                    }
+                }));
+            }
+        }
+        catch(const std::bad_alloc&)
+        {
+            return E_OUTOFMEMORY;
+        }
+
+        return S_OK;
+    }
+
+    void Terminate() noexcept
+    {
+        m_terminate = true;
+        m_wake.notify_all();
+
+        for(auto &thread : m_pool)
+        {
+            thread.join();
+        }
+
+        m_pool.clear();
+    }
+
+    void Submit() noexcept
+    {
+        std::lock_guard<std::mutex> lock(m_wakeLock);
```

Comment:
for reduced lock contention you may want to release lock before calling notify_all